pkgs/crane-utils/src/bin/crane-resolve-workspace-inheritance.rs

```diff
@@ -105,90 +96,300 @@ fn merge(cargo_toml: &mut toml::Value, root: &toml::Value) {
-        let (t, root_val) = match (&mut *v, root.get(k)) {
-            (toml::Value::Table(t), Some(root_val)) => (t, root_val),
-            _ => return,
-        };
-
-        if let Some(toml::Value::Boolean(true)) = t.get("workspace") {
-            t.remove("workspace");
-            let orig_val = mem::replace(v, root_val.clone());
-            merge_into(v, orig_val);
+        let (t, (root_key, root_val)) =
+            match try_as_table_like_mut(&mut *v).zip(root.get_key_value(&k)) {
+                Some((t, root_key_val)) => (t, root_key_val),
+                _ => return,
+            };
+
+        // copy any missing formatting to the current key
+        merge_key_formatting(k, root_key);
+
+        if let Some(Item::Value(toml_edit::Value::Boolean(bool_value))) = t.get("workspace") {
+            if *bool_value.value() {
+                t.remove("workspace");
+                let orig_val = mem::replace(v, root_val.clone());
+                merge_items(v, orig_val);
+            }
         }
     });
 }
 
-fn merge_into(dest: &mut toml::Value, additional: toml::Value) {
+/// Recursively merge the `additional` item into the specified `dest`
+fn merge_items(dest: &mut Item, additional: Item) {
+    use toml_edit::Value;
+
     match additional {
-        toml::Value::String(_)
-        | toml::Value::Integer(_)
-        | toml::Value::Float(_)
-        | toml::Value::Boolean(_)
-        | toml::Value::Datetime(_) => {
-            // Override dest completely for raw values
-            *dest = additional;
-        }
+        Item::Value(additional) => match additional {
+            Value::String(_)
+            | Value::Integer(_)
+            | Value::Float(_)
+            | Value::Boolean(_)
+            | Value::Datetime(_) => {
+                // Override dest completely for raw values
+                *dest = Item::Value(additional);
+            }
 
-        toml::Value::Array(additional) => {
-            if let toml::Value::Array(dest) = dest {
+            Value::Array(additional) => {
+                if let Item::Value(Value::Array(dest)) = dest {
+                    dest.extend(additional);
+                } else {
+                    // Override dest completely if types don't match
+                    *dest = Item::Value(Value::Array(additional));
+                }
+            }
+
+            Value::InlineTable(additional) => {
+                merge_tables(dest, additional);
+            }
+        },
+        Item::Table(additional) => {
+            merge_tables(dest, additional);
+        }
+        Item::None => {}
+        Item::ArrayOfTables(additional) => {
+            if let Item::ArrayOfTables(dest) = dest {
                 dest.extend(additional);
             } else {
-                // Override dest completely if types don't match
-                *dest = toml::Value::Array(additional);
+                *dest = Item::ArrayOfTables(additional);
```

Comment:
```suggestion
                 // Override dest completely if types don't match
                *dest = Item::ArrayOfTables(additional);
```

Just for consistency/clarity

Reply:
Updated, thanks! :+1: